lib/devicetrust/challenge/challenge.go

```diff
@@ -0,0 +1,103 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package challenge
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/rand"
+	"crypto/rsa"
+	_ "crypto/sha256" // imported from crypto.SHA256
+	"errors"
+	"fmt"
+
+	"github.com/gravitational/trace"
+)
+
+const challengeLength = 32
+
+// New creates a new Device Trust challenge with the default length.
+func New() ([]byte, error) {
+	b := make([]byte, challengeLength)
+	if _, err := rand.Read(b); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return b, nil
+}
+
+// Verify verifies that [sig] is a signature over [chal] by the private key
+// corresponding to [pubKey].
+func Verify(chal, sig []byte, pubKey crypto.PublicKey) error {
```

Comment:
Is this moved from e/? I think having challenge.Sign here makes sense, but I'm not sure about New or Verify on OSS.

We can do something like this on the e/challenge package to avoid importing 2 challenge packages server-side, if that's what we are trying to solve here:

```go
import challengeoss "github.com/gravitational/teleport/lib/devicetrust/challenge"

// Sign (...)
var Sign = challengeoss.Sign
```

Reply:
Actually we'll likely want Verify and Sign for testing, so I'll leave this all up to you.

Reply:
Yes I'm using both in tests, decided to just move the whole package out to OSS, nothing too special going on here imo

Reply:
I typically prefer e/ logic to be in e/ as to avoid multiple PRs for a single change, but like you said this is one of the safer/more stable packages to move.